geoips/schema/plugin_validator.py

```diff
@@ -33,60 +59,162 @@ def set_defaults(validator, properties, instance, schema):
-        schema_name = (
-            schema_file.replace(str(schema_path) + "/", "")
-            .replace("/", ".")
-            .replace(".yaml", "")
-        )
-        schema = yaml.safe_load(open(schema_file, "r"))
-        print(f"Adding validator {schema_name}")
+        schema_id = schema["$id"]
+
+        try:
+            validator.check_schema(schema)
+        except SchemaError as err:
+            raise SchemaError(f"Invalid schema {schema_file}") from err
+        schemas[schema_id] = schema
+
+    return schemas
 
-        DefaultValidatingValidator.check_schema(schema)
-        all_schema[schema_name] = schema
 
-    return all_schema
+def get_validators(schema_dict, validator_class):
+    """Create validators for each schema in `schema_dict`.
 
+    Parameters
+    ----------
+    schema_dict : dict
+        A dictionary whose keys are schema `$id`s and whose values are the full
+        schema.
 
-def get_validators(schema_dict):
+    Returns
+    -------
+    dict
+        A dictionary whose keys are schema `$id`s and whose values are `jsonschema`
+        validator instances.
+    """
+    # This is the jsonschema draft implemention from the referencing package
+    # It is used for creating the "resources" and "registry" that link together
+    # the references.
+    #
+    # This could likely have a better variable name, but I don't really know what to
+    # call it...
+    ref = getattr(refjs, f"DRAFT{JSONSCHEMA_DRAFT}")
+    resources = [(name, ref.create_resource(sch)) for name, sch in schema_dict.items()]
+    registry = referencing.Registry().with_resources(resources)
+
+    # Loop over the available schema and build a dictionary whose keys are the $ids
+    # of the schema and whose values are their associated validators. These validators
+    # can be used to validate yaml against their schema.
     validators = {}
     for name, schema in schema_dict.items():
-        resolver = jsonschema.RefResolver.from_schema(schema, store=schema_dict)
-        validators[name] = DefaultValidatingValidator(schema, resolver=resolver)
+        validators[name] = validator_class(schema, registry=registry)
     return validators
 
 
-all_schema = get_all_schema()
-validators = get_validators(all_schema)
-
-
-def validate(plugin_file):
-    """Validate the a YAML-based plugin."""
-    plugin_yaml = yaml.safe_load(open(plugin_file, "r"))
-    family = plugin_yaml["metadata"]["family"]
-    validator_name = plugin_yaml["interface"]
-    print(f"Checking validator {validator_name}.{family}")
-    if f"{validator_name}.{family}" in validators:
-        validator_name = f"{validator_name}.{family}"
-    print(f"Using validator {validator_name}")
-    validator = validators[validator_name]
-    validator.validate(plugin_yaml)
-
-    return plugin_yaml
+# schemas = get_schemas()
+# validators = get_validators(schemas)
+
+
+# def merge_product_defaults(plugin_yaml):
+#     for product in plugin_yaml["products"]:
+
+
+# def validate(plugin):
```

Comment:
Should we just delete this at this point?

Reply:
I removed the comments. Thanks!